packages/core/src/actions.ts

```diff
@@ -271,6 +272,39 @@ export function sendParent<
   });
 }
 
+type InferEvent<E extends EventObject> = {
+  [T in E['type']]: { type: T } & Extract<E, { type: T }>;
+}[E['type']];
```

Comment:
I am somewhat worried about the performance implications of this, but don't know how to make the whole thing work without it so 🤷‍♂️ I also might worry too much - but I definitely experience performance problems related to types when working with XState. It's hard to profile this though so... 🤷‍♂️